GA App Dependencies feature flag

mkangia opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
Make app dependencies feature flag available to Advanced plan or higher. All domains that have this enabled so far are on advanced plan or are test/demo projects so we don't need to specifically support feature flagged domains & can hence, deprecate the feature flag.

Deprecated feature flag can be removed in a follow up PR post deploy of this.
PR to be merged only after a go ahead from necessary division(s) and once comms are ready/done, Hence added label "do not merge".

A bugfix is also included in this PR ([bug](https://dimagi.atlassian.net/browse/QA-7196)).
App dependencies were not being revoked if the project didn't have access to the feature anymore so it continued to work for newer app versions as well.
This PR hence, if the project does not have access to the feature,
1. ignores app dependencies when generating the app profile for installation on the phone
2. wipes out any previous app dependencies set during saving of new profile settings on the app
The former seems good to have though the latter could be considered overkill. I am open to keeping just the former.

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SC-3942
Going on same lines as https://github.com/dimagi/commcare-hq/pull/33933

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally, tests and QA.
The changes done that impact domains outside the new privilege added, to revoke app dependencies if not accessible has been done in the safest way possible.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Tests failed on changes that were updated accordingly.
Mostly test classes were updated with mock to enable privilege by default so all tests navigate longest code paths, though for some the patch didn't work and hence were updated to use `TestCase` instead of `SimpleTestCase` to unblock tests and then patched separately where needed.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi.atlassian.net/browse/QA-7176

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [X] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

Reverting this PR will not cause any rollback on the records created/updated. Those extra records in the DB, will NOT conflict with the code deployed after the revert.
In case a permanent rollback of this PR is needed, then a new migration would be added to clean the redundant records so it can be done on all environments properly and not done manually.

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

mkangia committed:
refactor: align classes better

mkangia committed:
refactor: add kwargs for clarity

mkangia committed:
add new privilege

mkangia committed:
updates subscriptions with new privilege

mkangia committed:
updates checks for feature access

mkangia committed:
fixes other tests

mkangia committed:
deprecate toggle

mkangia commented:
Hey @Charl1996 
Yes, that is correct.

Charl1996 reviewed (approved):
LGTM

ajeety4 commented on corehq/apps/app_manager/tests/test_suite.py:
```diff
@@ -89,6 +92,7 @@ def test_owner_name(self, *args):
     def test_printing(self, *args):
         self._test_generic_suite('app_print_detail', 'suite-print-detail')
 
+    @privilege_enabled(privileges.APP_DEPENDENCIES)
```
Curious why this was required ?

mkangia replied:
Reasonable question @ajeety4 and well spotted.

Actually this test (and other ones) have been written using `SimpleTestCase` class which prohibits or is meant to not run DB queries.
The change I made for the feature, to check on privilege, fires up SQL queries, which caused tests to fail.

So, the option was to either change `SimpleTestCase` to `TestCase` or to mock the privilege check. I chose the latter to keep test db-independent and also because I would have done the patch anyway to let the test use app dependency feature by default.

mkangia replied:
Looking at the change 

https://github.com/dimagi/commcare-hq/pull/35237/files#diff-db20e55a79b6ed9c4bfbaa81acc7318970ec4634132220838b7abadb52379899R558

the tests do appear to have run with the toggle check. I understand that feature flags are also in DB but that still worked fine due to some unknown reason which I am okay to not dive into right now.

ajeety4 replied:
Noted. Thanks Manish for the clarification.

ajeety4 reviewed (commented):
Overall looks good. Just a non blocking query.

mkangia committed:
Merge branch 'master' into mk/3942-ga-app-dependencies

mkangia committed:
fix bad import line

mkangia commented:
Thanks for the prompt follow up on review @ajeety4

mkangia committed:
skip dependencies when missing privilege

mkangia committed:
revisit test_profile test fixes

- the privilege_enabled patch failed to work for these tests so resorted to using TestCase
- for one test that checks for profile, let it retain app dependencies

mkangia committed:
remove dependencies added on save

mkangia committed:
fix tests

mkangia committed:
remove unused import

mkangia commented:
Hey @ajeety4 @Charl1996 

I ended up making some fixes after your review.
Details about the bug are noted in the PR description under "Product description"

Charl1996 commented on corehq/apps/app_manager/views/settings.py:
```diff
@@ -73,6 +74,12 @@ def edit_commcare_profile(request, domain, app_id):
+
+    if not domain_has_privilege(domain, privileges.APP_DEPENDENCIES):
+        # remove dependencies if they were set before
+        if 'dependencies' in app.profile.get('features', {}):
```
Non blocking nit: because we're doing the same in [this commit](https://github.com/dimagi/commcare-hq/pull/35237/commits/791ada683322e0253e2190221a065d728873bec8), maybe create a separate function for doing this?

mkangia replied:
Oh yes that is the same thing at both places.

I believe I would be extracting just this bit
```
if 'dependencies' in app.profile.get('features', {}):
    del app.profile['features']['dependencies']
```

That seems a bit overkill?

Charl1996 replied:
Although it's only 2 lines, it's the concept that I was thinking of extracting, but I'll leave it to you to decide. I don't have strong feelings either way.

mkangia replied:
I agree with the concept but going to leave this as it is here.

ajeety4 commented:
Hey @mkangia ,
The new code changes looks good.

> App dependencies were not being revoked if the project didn't have access to the feature anymore so it continued to work for newer app versions as well.

Is it correct to say that this scenario won't occur in production as all the projects having the FF enabled are on the Plan to which the privilege is added ?
I am not well versed with this functionality and wanted to check if this fix added would be permanently needed ?

mkangia commented:
Hey @ajeety4 

> Is it correct to say that this scenario won't occur in production as all the projects having the FF enabled are on the Plan to which the privilege is added ?

That is correct for now for production environments. This scenario currently surfaced on staging during QA.
This could happen on production in future, if the subscription of a project is downgraded, and they lose access to the feature.

mkangia commented:
Comms [confirmed](https://dimagi.atlassian.net/browse/SC-3833?focusedCommentId=352464) so removing the blocking label.

mkangia commented:
Did some basic testing again locally to be confident about the changes considering the blast radius.

All went well so going to merge this now.
